kernel/sync_block/src/lib.rs

```diff
@@ -47,21 +66,55 @@ impl MutexFlavor for Block {
     where
         T: ?Sized,
     {
+        // Fast path
         // This must be a strong compare exchange, otherwise we could block ourselves
         // when the mutex is unlocked and never be unblocked.
         if let Some(guards) = Self::try_lock(mutex, data) {
-            guards
+            return guards;
+        }
+
+        // Middle path: Wait for one timeslice (see below) of the mutex holder to see if
+        // they release the lock.
+        let holder_id = data.holder.load(Ordering::Acquire);
+        if let Some(holder_task) = task::get_task(holder_id).and_then(|task| task.upgrade()) {
+            // Hypothetically, if holder_task is running on another core and is perfectly in
+            // sync with us, we would only ever check if they are running when we are also
+            // running and so we wouldn't detect when their timeslice is over. However, the
+            // likelihood of this is infinitesimally small and the code, is still correct as
+            // once the lock is released the holder will still set data.holder to 0 and we
+            // will exit the loop.
+            while holder_task.running_on_cpu().is_some()
+                && data.holder.load(Ordering::Acquire) == holder_id
+            {
+                core::hint::spin_loop();
+            }
+            if let Some(guards) = Self::try_lock(mutex, data) {
+                return guards;
+            }
         } else {
-            data.wait_until(|| Self::try_lock(mutex, data))
+            // Unlikely case that another thread just acquired the lock, but hasn't yet set
+            // data.holder.
+            log::warn!("could not get holder task for mutex middle path");
         }
+
```

Comment:
shall we try the fast path again here before blocking this task, considering that we just executed a series of "long-running" operations above as part of the middle path? I think so.

In other words, the lock may have been released whilst we were running `data.holder.load()` and `task::get_task(holder_id).and_then(|task| task.upgrade())`, so we can try to grab it before sleeping here.